src/test/java/MainTest.java

```diff
@@ -1,14 +1,94 @@
+import org.junit.Assert;
 import org.junit.Test;
 import org.jvnet.hudson.test.For;
+import org.jvnet.hudson.test.Issue;
 
+import javax.annotation.CheckForNull;
 import java.io.IOException;
 import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 @For(Main.class)
 public class MainTest {
 
+    private static final Logger LOGGER = Logger.getLogger(MainTest.class.getName());
+
     @Test(expected = IOException.class)
     public void shouldHaveNoStandardDependenciesFile() throws IOException {
         final Map<String, String> versions = Main.parseDependencyVersions();
     }
+
+    @Test
+    public void shouldFailForOldJava() {
+        assertJavaCheckFails(51, false);
+        assertJavaCheckFails(51, true);
+    }
+
+    @Test
+    public void shouldBeOkForJava8() {
+        assertJavaCheckPasses(52, false);
+        assertJavaCheckPasses(52, true);
+    }
+
+    @Test
+    public void shouldFailForMidJavaVersionsIfNoFlag() {
+        assertJavaCheckFails(53, false);
+        assertJavaCheckFails(54, false);
+        assertJavaCheckPasses(53, true);
+        assertJavaCheckPasses(54, true);
+    }
+
+    @Test
+    @Issue("JENKINS-51805")
+    public void shouldBeOkForJava11() {
+        assertJavaCheckPasses(55, false);
+        assertJavaCheckPasses(55, true);
+    }
+
+    @Test
+    public void shouldFailForNewJavaVersionsIfNoFlag() {
+        assertJavaCheckFails(56, false);
+        assertJavaCheckFails(57, false);
+        assertJavaCheckPasses(56, true);
+        assertJavaCheckPasses(57, true);
+    }
+
+    public void assertJavaCheckFails(int classVersion, boolean enableFutureJava) {
+        assertJavaCheckFails(null, classVersion, enableFutureJava);
+    }
+
+    public void assertJavaCheckFails(@CheckForNull String message, int classVersion, boolean enableFutureJava) {
+        boolean failed = false;
+        try {
+            Main.verifyJavaVersion(classVersion, enableFutureJava);
+        } catch (Error error) {
+            failed = true;
+            LOGGER.log(Level.WARNING, "Java class version check failed as it was expected", error);
```

Comment:
Oh, and I guess putting a warning in tests is un-great, as it is going to spit a lot of warnings to the console for the nominal case, but oh well.

Reply:
JUL works well with the default log appenders, except some race conditions. I can switch to STDOUT/STDERR if you prefer